internal/service/alarms/internal/alarms_server.go

```diff
@@ -363,11 +364,43 @@ func (a *AlarmsServer) GetProbableCause(ctx context.Context, request api.GetProb
 	return nil, fmt.Errorf("not implemented")
 }
 
+// AmNotification handles an API request coming from AlertManager with CaaS alerts. This api is used internally.
+// Note: the errors returned can also be view under alertmanager pod logs but also logging here for convenience
 func (a *AlarmsServer) AmNotification(ctx context.Context, request api.AmNotificationRequestObject) (api.AmNotificationResponseObject, error) {
-	// TODO: Implement the logic to handle the AM notification
-	slog.Debug("Received AM notification", "groupLabels", request.Body.GroupLabels)
-	for _, alert := range request.Body.Alerts {
-		slog.Debug("Alert", "fingerprint", alert.Fingerprint, "startsAt", alert.StartsAt, "status", alert.Status)
+	// Audit the table with full list of alerts in the current payload. If missing set them to resolve
+	if err := a.AlarmsRepository.ResolveNotificationIfNotInCurrent(ctx, request.Body); err != nil {
+		msg := "failed to resolve notification that are not present"
+		slog.Error(msg, "error", err)
+		return nil, fmt.Errorf("%s: %w", msg, err)
+	}
+
+	// Get the definition data based on current set of Alert names and managed cluster ID
+	alarmDefinitions, err := a.AlarmsRepository.GetAlarmDefinitions(ctx, request.Body, a.ClusterServer.ClusterIDToResourceTypeID)
+	if err != nil {
+		msg := "failed to get AlarmDefinitions"
+		slog.Error(msg, "error", err)
+		return nil, fmt.Errorf("%s: %w", msg, err)
+	}
+
+	// Combine possible definitions with events
+	aerModels := alertmanager.ConvertAmToAlarmEventRecordModels(request.Body, alarmDefinitions, a.ClusterServer.ClusterIDToResourceTypeID)
+
+	// Insert and update AlarmEventRecord
+	if err := a.AlarmsRepository.UpsertAlarmEventRecord(ctx, aerModels); err != nil {
+		msg := "failed to upsert AlarmEventRecord to db"
+		slog.Error(msg, "error", err)
+		return nil, fmt.Errorf("%s: %w", msg, err)
+	}
+
+	//TODO: Get subscriber
+
+	//TODO: Notify subscriber
+
+	// Removed resolved events and move to archive
+	if err := a.AlarmsRepository.ArchiveResolvedAlarmEventRecords(ctx); err != nil {
```

Comment:
Why do we need to archive the resolved alarms? At first, I thought this table was to archive the alerts that were purged. A place to have an extra backup for users to check. But you are moving everything that has been resolved. 

The Alarm Records Endpoints use the regular alarm_record table. There must be resolved and firing alarms for the users to query whatever they want using filters. With this, they won't be able to query resolved alerts. They should, these are supposed to be gone only after the retention period has expired.

Reply:
There is a requirement to have an historical alarm DB

Reply:
I understand that but that doesn't mean we should delete the resolved ones from the main Alarm Record table. Not before the retention period expires.

So, this method `ArchiveResolvedAlarmEventRecords` is deleting the resolved alarms from the main table `alarm_event_record` and then moving those to the archive table. This means that the main table will only have unresolved alarms. Our Alarm Record enpoints are using only the main alarm record table so queries for resolved alerts will be empty.

The job of archiving the resolved alerts should be done by the k8s Job we will implement in charge of purging the resolved alerts from the main table. Only then, they become purged and we could safely archive them without affecting expected workflow.

Another thing, the specification says that we need to purge resolved alarm and also active acknowledged alerts.

Reply:
ok, so for the resolved alarms they should not be moved until retention period expires. 
I assume we would ignore these on a resync with the alert server ? 

As for the purge description in the spec. 
We cannot purge active acknowledged alarms as they still exist in the alert server, we would just readd on the next sync.

Reply:
Correct. That's why I'm suggesting @pixelsoccupied  to identify which Alerts are already in the DB and simply ignore them instead of upserting the rows. If we are assuming that it is safe to resolve records that are not in the reported list from the Alert server (AlertManager), we could also assume that the ones that are present in the list and not in the DB are new ones. We could work only with these ones.

Regarding the active acknowledged alarms, that's a challenged indeed. Perhaps, we could also check the archive table to avoid the re triggering since you are mentioning, we don't control the Alert server (AlertManager). Those alerts will still be active. Or like you are suggesting, simply not purge them.

Reply:
yeah basically if the user provides a filter to look for archived/resolved data we would simply look into the archive table...otherwise we simply check the active data. 

I think original assumption behind this is we will keep one regular table with active events and the rest goes into the archive so that active table doesnt get too large. And archive table is what we delete at an internal. 

as an optimization we can also leverage DB so that they both appear as the same table such as union view (tbh since i have looked into the filter yet)

Reply:
Why do we have to overdesign this? Let's keep it simple. This endpoint does not archive anything. The purge Job will move resolved alerts to the archive table when the retention period expires. With this, Alert Record endpoints will remain simple and use the main table. No tricks with multiple DBs. The archive table can have its own rentention period and its owns mean to serve queries (outside of the specification)

What do you think? @pixelsoccupied @browsell

Reply:
deleted